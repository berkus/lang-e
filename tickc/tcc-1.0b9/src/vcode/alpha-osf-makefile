CC=gcc
BPOCFLAGS=-D__no_bpo__
CFLAGS= -Wall -g $(BPOCFLAGS)
LIBS=
VARCH=alpha
TST=tst
HDRS=vcode.h vcode-macros.h
VOBJS= link.o $(VARCH)-gen.o gen.o register.o xlate.o call.o poly-insns.o fp-imm.o alpha-flushcache.o portable-insts.o mult.o
VSRC= $(VOBJS:.o=.c)

all: binary.h vcode-macros.h vcode.lib.a

all: vcode.lib.a 

vcode.lib.a: $(VOBJS) $(HDRS)
	ar ruv vcode.lib.a alpha-dis/* $(VOBJS)
	ranlib vcode.lib.a

binary.h: $(VARCH)-bin
	./bpp/bpp < $(VARCH)-bin > binary.h

vcode-macros.h: binary.h $(VARCH).md spec.pl
	perl spec.pl fast < $(VARCH).md > vcode-macros.h

$(VARCH)-gen.c: $(HDRS)
gen.c: $(HDRS)

# code for cmul
CMUL = 4096
mult.h: booth-gen
	./booth-gen $(CMUL) > mult.h
booth-gen: booth-gen.c
	$(CC) $(CFLAGS) -o booth-gen booth-gen.c

depend:
	makedepend $(VSRC)
clean:
	rm -f *.o core vcode.lib.a *.Counts *.pixie *.Addrs vcode-macros.h binary.h
# DO NOT DELETE THIS LINE -- make depend depends on it.

link.o: vcode.h vcode-config.h vcode-macros.h demand.h /usr/include/stdio.h
link.o: /usr/include/standards.h binary.h /usr/include/assert.h
alpha-gen.o: /usr/include/assert.h /usr/include/standards.h
alpha-gen.o: /usr/include/stdio.h /usr/include/stdarg.h /usr/include/string.h
alpha-gen.o: /usr/include/sys/types.h demand.h vcode-internal.h vcode.h
alpha-gen.o: vcode-config.h vcode-macros.h binary.h
gen.o: /usr/include/assert.h /usr/include/standards.h /usr/include/stdio.h
gen.o: vcode-internal.h vcode.h vcode-config.h vcode-macros.h demand.h
gen.o: binary.h
register.o: vcode-internal.h vcode.h vcode-config.h vcode-macros.h demand.h
register.o: /usr/include/stdio.h /usr/include/standards.h binary.h
register.o: /usr/include/assert.h /usr/include/string.h
register.o: /usr/include/sys/types.h
xlate.o: vcode-internal.h vcode.h vcode-config.h vcode-macros.h demand.h
xlate.o: /usr/include/stdio.h /usr/include/standards.h binary.h
xlate.o: /usr/include/assert.h
call.o: /usr/include/stdio.h /usr/include/standards.h vcode-internal.h
call.o: vcode.h vcode-config.h vcode-macros.h demand.h binary.h
call.o: /usr/include/assert.h
poly-insns.o: vcode.h vcode-config.h vcode-macros.h demand.h
poly-insns.o: /usr/include/stdio.h /usr/include/standards.h binary.h
poly-insns.o: /usr/include/assert.h
fp-imm.o: vcode-internal.h vcode.h vcode-config.h vcode-macros.h demand.h
fp-imm.o: /usr/include/stdio.h /usr/include/standards.h binary.h
fp-imm.o: /usr/include/assert.h
mult.o: vcode.h mult.h
