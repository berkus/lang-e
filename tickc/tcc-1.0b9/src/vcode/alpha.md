#include "demand.h"
#include "binary.h"

extern void v_jalpi(v_reg_type r, void *ip);
extern void cvd2l(int rd, int rs);
extern void cvd2i(int rd, int rs);
extern void cvl2f(int rd, int rs);
extern void cvl2d(int rd, int rs);

/* Need to fix the damn preprocessor. */
#define v_alduci(rd, rs, offset, align)	\
	alduc(_vrr(rd), _vrr(rs), offset, align)

#define v_aldci(rd, rs, offset, align)	\
	aldc(_vrr(rd), _vrr(rs), offset, align)

#define v_aldsi(rd, rs, offset, align)	\
	alds(_vrr(rd), _vrr(rs), offset, align)

#define v_aldusi(rd, rs, offset, align)	\
	aldus(_vrr(rd), _vrr(rs), offset, align)

#define v_aldui(rd, rs, offset, align)	\
	aldl(_vrr(rd), _vrr(rs), offset, align)
#define v_aldi	v_aldui

#define v_aldli(rd, rs, offset, align)	\
	aldq(_vrr(rd), _vrr(rs), offset, align)
#define v_alduli	v_aldli
#define v_aldpi	v_aldli

/* Only do a jump-subroutine when we are actually going to do a call */
#define v_jalp(r, dst) do {				\
	if(_vrr(r) == _ra) {				\
	     jsr(_vrr(r), _vrr(dst), 0);		\
	} else {					\
		jmp_link(_vrr(r), _vrr(dst)); 		\
	}						\
} while(0)

#define v_jpi(dst)       br(_zero, (unsigned)(dst))
#define v_jp(dst)        jmp_nolink(_vrr(dst))
/* Inverted order of jpi and jmark to make bpo work correctly */
#define v_jv(l)		do { v_jpi(0); v_jmark(v_ip, l); } while(0)
#define v_retv() 	v_jv(v_epilogue_l)

#define v_setp(r, imm)	v_setul(r, (unsigned long)imm)

#define v_nop() nop()

extern v_label_type v_epilogue_l;
extern void stus(int rd, int rs, long offset);
extern void ldus(int rd, int rs, long offset);

/* v_schedule_delay: alpha has no delay slots.  Place in order */
#define v_schedule_delay(branch, insn) do {  insn; branch; } while(0)

/* v_raw_load: alpha does not have delay slots. */
#define v_raw_load(mem, n) mem
#define v_nuke_nop()

extern void pseudo_call(v_vptr insn, int rd, int rs1, int rs2);

#define v_setf(rd, imm) v_float_imm(rd, imm)
#define v_setd(rd, imm) v_double_imm(rd, imm)

#define v_retfi(imm) do { v_setf(v_fp_rr, imm); v_retv(); } while(0)
#define v_retdi(imm) do { v_setd(v_fp_rr, imm); v_retv(); } while(0)

%%
(ret ((@ v_callee_int_rr, rs_or_imm) (v_jv v_epilogue_l)) (i l u ul p  v_mov: v_set:))
(ret ((@ v_fp_rr, rs) (v_jv v_epilogue_l)) (f d v_mov:))

(add (i u addl addli) (l ul p addq addqi) (f adds) (d addt))
(sub (i u subl subli) (l ul p subq subqi) (f subs) (d subt))

(mul (s i u mull mulli) (l ul mulq mulqi) (f muls) (d mult))

; The great thing about alpha pseudo-insns is that they 
; save temp registers.

(div (s i ((pseudo_call __divl, rd, rs1, rs2)))
	(u ((pseudo_call __divlu, rd, rs1, rs2)))
	(l ((pseudo_call __divq, rd, rs1, rs2)))
	(ul ((pseudo_call __divqu, rd, rs1, rs2)))
	(f divs) (d divt))
(mod (s i ((pseudo_call __reml, rd, rs1, rs2)))
	(u ((pseudo_call __remlu, rd, rs1, rs2)))
	(l ((pseudo_call __remq, rd, rs1, rs2)))
	(ul ((pseudo_call __remqu, rd, rs1, rs2))))

(def no_reg_plus_reg)
(ld (c uldc) (uc ulduc) (s ulds) (us uldus) 
	(i u ldl)
	(ul l p ldq)
	(f lds) (d ldt))
(st (c uc ustb) (s us usts)
	(i u stl)
	(ul l p stq) 
	(f sts) (d stt))

(uld (s ulds) (us uldus) 
	(i u uldl)
	(ul l p uldq))

(ust (s us usts)
	(i u ustl)
	(ul l p ustq) )

; sort of boolean ops
(lsh
        (i
		; sign extension
                ((sll rd, rs1, rs2) (addl rd, _zero, rd))
                ((slli rd, rs1, imm) (addl rd, _zero, rd)))
        (u
                ((zapnoti rd, rs1, 0xf) (sll rd, rd, rs2))
                ((zapnoti rd, rs1, 0xf) (slli rd, rd, imm)))
        (l ul sll slli))

(rsh (i l sra srai) (ul srl srli)
	(u 
		((zapnoti rd, rs1, 0xf) (srl rd, rd, rs2))
		((zapnoti rd, rs1, 0xf) (srli rd, rd, imm))))

(and (us s i l u ul and andi))
(or (us s i l u ul bis bisi))
(xor (us s i l u ul xor xori))

(nor (i u l ul ( (v_or: rd, rs1, rs2) (v_com: rd, rd) ) ))
(nxor (i u l ul ((v_xor: rd, rs1, rs2) (v_com: rd, rd))))
(nand (i u l ul ((v_and: rd, rs1, rs2) (v_com: rd, rd))))

(andnot (i u l ul andnot andnoti))
(ornot (i u l ul ornot ornoti))
(xornot (i u l ul xornot xornoti))

(com (i l u ul com))

(neg 
	(i u ((subl rd, _zero, rs))) 
	(l ul ((subq rd, _zero, rs)))
  	(f d fneg))
(not (i l u ul not))
(mov (i l u ul p mov) (f d fmov))

(set (c uc s us i u l set) (ul setu))

(abs (f fabss) (d fabsd))
(nabs (f d ((v_abs: rd, rs) (v_neg: rd, rd))))
(sqrt (f fsqrts) (d fsqrtd))

; some useful macros
(def sext24 ((v_lshui rd,rs,24) (v_rshii rd,rd,24)))
(def sext16 ((v_lshui rd,rs,16) (v_rshii rd,rd,16)))
(def trunc16 ((v_andii rd,rd,0xffff)))
; Uch
; (def f2l ((fstoi rs, rs) (stf rs, v_lpr, v_carea) (ld rd, v_lpr, v_carea)))
; (def d2l ((fdtoi rs, rs) (stf rs, v_lpr, v_carea) (ld rd, v_lpr, v_carea)))
; (def i2f ((st rs, v_lpr, v_carea) (ldf rd, v_lpr, v_carea)))
; (def i2d ((std rs, v_lpr, v_carea) (lddf rd, v_lpr, v_carea)))

(cvc2 (i u l ul sext24))
(cvs2 (i u l ul sext16))
(cvus2 (i u l ul trunc16))
(cvi2 (c sext24) (s sext16) (u l ul mov))
(cvu2 (c sext24) (s sext16) (i mov) (l ul ((zapnoti rd, rs, 0xf))))
(cvl2 (c sext24) (s sext16) (i u ((addl rd, rs, _zero))) (ul mov) (f cvl2f) (d cvl2d))
(cvul2 (c sext24) (s sext16) (u i ((addl rd, rs, _zero))) (l p mov))
(cvp2 (ul mov))	
(cvf2 (d ((addt rd, rs, _fzero))) (l cvd2l))
(cvd2 (f ((cvtts rd, rs))) (l cvd2l))

(blt (i l cmplq cmplqi) (u cmpluq cmpluqi) (ul p cmpluq cmplulqi) (f d bltf))
(bge (i l cmpgeq cmpgeqi) (u cmpgeuq cmpgeuqi) (ul p cmpgeuq cmpgeulqi) (f d bgef))
(ble (i l cmpleq cmpleqi) (u cmpleuq cmpleuqi) (ul p cmpleuq cmpleulqi) (f d blef))
(bgt (i l cmpgtq cmpgtqi) (u cmpgtuq cmpgtuqi) (ul p cmpgtuq cmpgtulqi) (f d bgtf))
(beq (i u cmpeql cmpeqli) (ul l p cmpeqq cmpeqqi) (f d beqf))
(bne (i u cmpnel cmpneli) (l ul p cmpneq cmpneqi) (f d bnef))

(muladd ((v_mul: v_dat, rs1, rs2) (v_add: rd, v_dat, rs3)) (f) (d))
(mulsub ((v_mul: v_dat, rs1, rs2) (v_sub: rd, v_dat, rs3)) (f) (d))
(negmuladd ((v_mul: v_dat, rs1, rs2) (v_add: rd, v_dat, rs3) (v_neg: rd,rd)) (f) (d))
(negmulsub ((v_mul: v_dat, rs1, rs2) (v_sub: rd, v_dat, rs3) (v_neg: rd,rd)) (f) (d))
