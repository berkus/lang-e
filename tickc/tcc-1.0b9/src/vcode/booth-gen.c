#include <stdio.h>
#include <stdlib.h>


#if (defined(SPARC) || defined(__sparc__)) && defined(__GNUC__) && !defined(_V_SOLARIS_) && !defined(__SVR4)
extern int printf(char *, ...);
#endif

#define ADD_COST 1
#define SUB_COST 1
#define NEG_COST 1
#define SHIFT_COST 1
#define HASH_SIZE 511

#undef verbose

static int      t, n, src;

typedef
enum {
	IDENTITY,		/* used for n = 1          */
	NEGATE,			/* used for n = -1         */
	SHIFT_ADD,		/* used for makeOdd(n - 1) */
	SHIFT_SUB,		/* used for makeOdd(n + 1) */
	SHIFT_REV,		/* used for makeOdd(1 - n) */
	FACTOR_ADD,		/* used for n/(2^i - 1)    */
	FACTOR_SUB,		/* used for n/(2^i + 1)    */
	FACTOR_REV		/* used for n/(1 - 2^i)    */
}               MulOp;

typedef
struct node {
	struct node    *parent;
	int             value;
	unsigned int    cost;
	MulOp           opcode;
	struct node    *next;
}               Node;

static unsigned int costs[] =
{0,				/* for IDENTITY   */
	NEG_COST,		/* for NEGATE     */
	SHIFT_COST + ADD_COST,	/* for SHIFT_ADD  */
	SHIFT_COST + SUB_COST,	/* for SHIFT_SUB  */
	SHIFT_COST + SUB_COST,	/* for SHIFT_REV  */
	SHIFT_COST + ADD_COST,	/* for FACTOR_ADD */
	SHIFT_COST + SUB_COST,	/* for FACTOR_SUB */
	SHIFT_COST + SUB_COST	/* for FACTOR_REV */
};

static Node    *hash_table[HASH_SIZE];

#define odd(c) ((c) & 1)
#define even(c) (!odd(c))

static int 
makeOdd(int c)
{
	do
		c = c / 2;
	while (even(c));
	return c;
}

static Node    *
lookup(int c)
{
	int             hash = abs(c) % HASH_SIZE;
	Node           *node = hash_table[hash];
	while (node && node->value != c)
		node = node->next;
	if (!node) {
		node = (Node *) malloc(sizeof(Node));
		node->value = c;
		node->parent = NULL;
		node->cost = SHIFT_COST + 1;

		node->next = hash_table[hash];
		hash_table[hash] = node;
	}
	return node;
}

static void 
emit_shift(int target, int source)
{
	int             temp = source;
	unsigned int    i = 0;
	do {
		temp <<= 1;
		i++;
	} while (target != temp);
#ifdef verbose
	fprintf(stdout, "%d = %d << %d\n", target, source, i);
#endif
	n = i;
}

static int 
emit_code(Node * node)
{
	int             source;
#if 0
	unsigned int    shift;
#endif
	int             target = node->value;
	switch (node->opcode) {
	case IDENTITY:
		break;
	case NEGATE:
		source = emit_code(node->parent);
		printf("\tNEGATE,");
		t = -t;
#ifdef verbose
		fprintf(stdout, "%d = 0 - %d\n", target, source);
#endif
		break;
	case SHIFT_ADD:
		source = emit_code(node->parent);
		emit_shift(target - 1, source);
#ifdef verbose
		fprintf(stdout, "%d = %d + 1\n", target, target - 1);
#endif
		printf("\tSHIFT_ADD, %d,", n);
		t = (t << n) + src;
		break;
	case SHIFT_SUB:
		source = emit_code(node->parent);
		emit_shift(target + 1, source);
#ifdef verbose
		fprintf(stdout, "%d = %d - 1\n", target, target + 1);
#endif
		printf("\tSHIFT_SUB, %d,",n);
		t = (t << n) - src;
		break;
	case SHIFT_REV:
		source = emit_code(node->parent);
		emit_shift(1 - target, source);
#ifdef verbose
		fprintf(stdout, "%d = 1 - %d\n", target, 1 - target);
#endif
		printf("\tSHIFT_REV, %d,",n);
		t = src - (t << n);
		break;
	case FACTOR_ADD:
		source = emit_code(node->parent);
		emit_shift(target - source, source);
#ifdef verbose
		fprintf(stdout, "%d = %d + %d\n", target, target - source, source);
#endif
		t = (t << n) + t;
		printf("\tFACTOR_ADD, %d,",n);
		break;
	case FACTOR_SUB:
		source = emit_code(node->parent);
		emit_shift(target + source, source);
#ifdef verbose
		fprintf(stdout, "%d = %d - %d\n", target, target + source, source);
#endif
		t = (t << n) - t;
		printf("\tFACTOR_SUB, %d,",n);
		break;
	case FACTOR_REV:
		source = emit_code(node->parent);
		emit_shift(source - target, source);
#ifdef verbose
		fprintf(stdout, "%d = %d - %d\n", target, source, source - target);
#endif
		t = t - (t << n);
		printf("\tFACTOR_REV, %d,",n);
		break;
	}
	return target;
}

static Node    *find_sequence(int c, unsigned int limit);

static void 
try(int factor, Node * node, MulOp opcode)
{
	unsigned int    cost = costs[opcode];
	unsigned int    limit = node->cost - cost;
	Node           *factor_node = find_sequence(factor, limit);
	if (factor_node->parent && factor_node->cost < limit) {
		node->parent = factor_node;
		node->opcode = opcode;
		node->cost = factor_node->cost + cost;
	}
}

static Node    *
find_sequence(int c, unsigned int limit)
{
	Node           *node = lookup(c);
	if (!node->parent && node->cost < limit) {
		node->cost = limit;
		if (c > 0) {
			int             power = 4;
			int             edge = c >> 1;
			while (power < edge) {
				if (c % (power - 1) == 0)
					try(c / (power - 1), node, FACTOR_SUB);
				if (c % (power + 1) == 0)
					try(c / (power + 1), node, FACTOR_ADD);
				power = power << 1;
			}
			try(makeOdd(c - 1), node, SHIFT_ADD);
			try(makeOdd(c + 1), node, SHIFT_SUB);
		} else {
			int             power = 4;
			int             edge = (-c) >> 1;
			while (power < edge) {
				if (c % (1 - power) == 0)
					try(c / (1 - power), node, FACTOR_REV);
				if (c % (power + 1) == 0)
					try(c / (power + 1), node, FACTOR_ADD);
				power = power << 1;
			}
			try(makeOdd(1 - c), node, SHIFT_REV);
			try(makeOdd(c + 1), node, SHIFT_SUB);
		}
	}
	return node;
}

static unsigned int 
estimate_cost(int c)
{
	/*
	 * if (c >= 0) if (c >= 65536) return 9; else return 5; else if (c <=
	 * -65536) return 9; else return 5;
	 */
	return 20;
}

void 
multiply(int target)
{
	unsigned int    multiply_cost = estimate_cost(target), cost;
	if (odd(target)) {
		Node           *result = find_sequence(target, multiply_cost);
		if (result->parent && (cost = result->cost) < multiply_cost) {
			printf("\t%d,",cost);
			(void) emit_code(result);
		}
#ifdef verbose
		else
		     fprintf(stdout, "use multiply instruction\n");
#endif
	} else {
		Node           *result = find_sequence(makeOdd(target), multiply_cost - SHIFT_COST);
		if (result->parent && (cost = (result->cost + SHIFT_COST)) < multiply_cost) {
			int             source;
			printf("\t%d,",cost);
			source = emit_code(result);
			emit_shift(target, source);
			t = t << n;
			printf("\tSHIFT, %d,",n);
		} 
#ifdef verbose
		else
		     	fprintf(stdout, "use multiply instruction\n");
#endif
	}
}

void 
init_multiply(void)
{
	Node           *node, *node1;
	unsigned int    i;
	for (i = 0; i < HASH_SIZE; i++)
		hash_table[i] = NULL;

	node1 = lookup(1);
	node1->parent = node1;	/* must not be NULL */
	node1->opcode = IDENTITY;
	node1->cost = 0;

	node = lookup(-1);
	node->parent = node1;
	node->opcode = NEGATE;
	node->cost = NEG_COST;
}

void 
main(int argc, char *argv[])
{
	int             i, scaler = atoi(argv[1]);
	init_multiply();

	printf(" enum {    \n"
	"\t	IDENTITY, /* used for n = 1          */			\n"
	"\t	NEGATE,			/* used for n = -1         */   \n"
	"\t	SHIFT_ADD,		/* used for makeOdd(n - 1) */   \n"
	"\t	SHIFT_SUB,		/* used for makeOdd(n + 1) */   \n"
	"\t	SHIFT_REV,		/* used for makeOdd(1 - n) */   \n"
	"\t	FACTOR_ADD,		/* used for n/(2^i - 1)    */   \n"
	"\t	FACTOR_SUB,		/* used for n/(2^i + 1)    */   \n"
	"\t	FACTOR_REV,		/* used for n/(1 - 2^i)    */   \n"
	"\t	SHIFT,							\n"
	"\t	DONE							\n"
	" };\n\n");


	for(i = 1; i <= scaler; i++) {
		printf("static char mul%d[] = {",i);
		multiply(i);
		printf("\tDONE };\t/* %d */\n",i);
	}
	printf("\n#define LSIZE %d\n",scaler+1);
	printf("\n\nstatic char *mlookup[LSIZE] = {\n\t0,\n");
	for(i = 1; i <= scaler; i++) {
		printf("\tmul%d,\n",i);			
	}
	printf("};\n");

#if 0
	while (EOF != scanf("%d", &i)) {
		src = t = 2;
		if (i)
			multiply(i);
		printf("%d * %d = %d,  t = %d\n", i, 2, i * 2, t);
	}
#endif

	exit(0);
}
