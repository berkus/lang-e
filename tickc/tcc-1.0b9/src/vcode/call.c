/***********************************************************************
 * Call management.
 */
#define V_MAKE_REG_MAP
#include <stdio.h>
#include "vcode-internal.h"
#include "demand.h"


/* Allocate a register argument. */
static void v_arg(struct v_cstate *c, int type, v_reg_type r) {
	struct v_arg a;

	a.isreg_p = 1;
	a.type = type;
	a.u.r = r;

	v_mach_push(c, type, &a);
}

/* Allocate an immediate argument: poor-man's polymorphism. */
#define ARGI(c, t, imm) do {				\
	struct v_arg a;					\
	a.isreg_p = 0;					\
	a.type = t;					\
	a.u.imm = imm;					\
	v_mach_push(c, t, &a);				\
} while(0)

/* 
 * The different argument types. We have seperated these routines
 * to allow better type checking on immediates.  Once they were
 * provided for immediates, it seemed unnatural to not provide them
 * for registers.
 */
void v_push_argi(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_I, r); }
void v_push_argii(struct v_cstate *c, int i) 	
	{ ARGI(c, V_I, i); }
void v_push_argu(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_U, r); }
void v_push_argui(struct v_cstate *c, unsigned u) 	
	{ ARGI(c, V_U, u); }
void v_push_argl(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_L, r); }
void v_push_argli(struct v_cstate *c, long l) 	
	{ ARGI(c, V_L, l); }
void v_push_argul(struct v_cstate *c, v_reg_type r)  
	{ v_arg(c, V_UL, r); }
void v_push_arguli(struct v_cstate *c, unsigned long ul) 
	{ ARGI(c, V_UL, ul); }
void v_push_argp(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_P, r); }
void v_push_argpi(struct v_cstate *c, void *p) 	
	{ ARGI(c, V_P, p); }
void v_push_argf(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_F, r); }
void v_push_argfi(struct v_cstate *c, float f) 	
	{ ARGI(c, V_F, f); }
void v_push_argd(struct v_cstate *c, v_reg_type r) 	
	{ v_arg(c, V_D, r); }
void v_push_argdi(struct v_cstate *c, double d) 	
	{ ARGI(c, V_D, d); }

/* generic push. */
void v_arg_push(struct v_cstate *c, int type, v_reg_type r) {
	v_arg(c, type, r);
}

static void ccall(struct v_cstate *c, void (*ptr)()) {
	v_mach_call(c, ptr);
}

inline void v_ccallv(struct v_cstate *c, v_vptr ptr) { 
	ccall(c, ptr); 
}

inline v_reg_type v_ccalli(struct v_cstate *c, v_iptr ptr) {
	ccall(c, (v_vptr)ptr);
	return v_caller_int_rr;
}

inline v_reg_type v_ccallu(struct v_cstate *c, v_uptr ptr) {
	return v_ccalli(c, (v_iptr)ptr);
}

inline v_reg_type v_ccallp(struct v_cstate *c, v_pptr ptr) {
	return v_ccalli(c, (v_iptr)ptr);
}

inline v_reg_type v_ccalll(struct v_cstate *c, v_lptr ptr) {
	return v_ccalli(c, (v_iptr)ptr);
}

inline v_reg_type v_ccallul(struct v_cstate *c, v_ulptr ptr) {
	return v_ccalli(c, (v_iptr)ptr);
}

inline v_reg_type v_ccallf(struct v_cstate *c, v_fptr ptr) {
	ccall(c, (v_vptr)ptr);
	return v_fp_rr;
}

inline v_reg_type v_ccalld(struct v_cstate *c, v_dptr ptr) {
	ccall(c, (v_vptr)ptr);
	return v_fp_rr;
}

/*
 * Machinery for simple calls.
 */

/* 
 * A bit of syntactic sugar: all scall does is pull the arguments off
 * of the stack and then call ccall. 
 */
static void scall(void (*ptr)(), char *fmt, va_list ap) {
	int nargs, i, *arglist;
	struct v_cstate c;

	if(!v_pseudo_call && v_isleaf)
		v_fatal("cannot do a call from a leaf!\n");


	v_calls++;
	arglist = v_xlatel(fmt, &nargs);
	
	if(nargs >= V_MAXARGS)
		v_fatal("scall: too many arguments. Maximum is %d, gave %d\n", 
					V_MAXARGS, nargs);
	v_push_init(&c);

	/* push all arguments */
	for(i = 0; i < nargs; i++) {
		if(!(arglist[i] & V_IMMEDIATE))
		  v_arg_push(&c, arglist[i], va_arg(ap, v_reg_type));
		else {
			/* mask out the immediate flag */
			switch(arglist[i] & ~V_IMMEDIATE) {
			case V_UC:
			case V_US:
			case V_U:   v_push_argui(&c, va_arg(ap, unsigned)); break;
			case V_C:
			case V_S:
			case V_I:   v_push_argii(&c, va_arg(ap, int)); break;
			case V_L:   v_push_argli(&c, va_arg(ap, long)); break;
			case V_UL:  v_push_arguli(&c, va_arg(ap, unsigned long)); break;
			case V_P:   v_push_argpi(&c, va_arg(ap, void *)); break;
			case V_F:   v_push_argfi(&c, (float)va_arg(ap, double)); break;
			case V_D:   v_push_argdi(&c, va_arg(ap, double)); break;
			default:    v_fatal("v_scall*: bogus type\n"); 
			}
		}
	}
	ccall(&c, ptr);
}


void v_scallv(v_vptr ptr, char *fmt, ...) {
        va_list ap;

        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
}

v_reg_type v_scalli(v_iptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_caller_int_rr;
}
v_reg_type v_scallu(v_uptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_caller_int_rr;
}

v_reg_type v_scallp(v_pptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_caller_int_rr;
}

v_reg_type v_scallul(v_ulptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_caller_int_rr;
}

v_reg_type v_scalll(v_lptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_caller_int_rr;
}
v_reg_type v_scallf(v_fptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_fp_rr;
}

v_reg_type v_scalld(v_dptr ptr, char *fmt, ...) {
        va_list ap;
        va_start(ap, fmt);
        scall((v_vptr)ptr, fmt, ap);
        va_end(ap);
        return v_fp_rr;
}

void v_fatal(char *fmt, ...) {
	va_list ap;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	exit(1);
}

/* Map a symbolic register name to a physical register name. */
v_reg_type v_sym_to_phys(char *r) {
	struct reg_map *rm;

	if(!r)
		v_fatal("v_sym_to_phys: register name must be non-nil.\n");
		
	for(rm = sym_to_phys; rm->sym_name; rm++)
		if(strcmp(rm->sym_name,  r) == 0)
			return v_reg(rm->phys_name);

	v_fatal("v_sym_to_phys: unrecognized register name %s\n", r);
	/*NOTREACHED*/
	return v_reg(0);	/* to shut up -Wall */
}

void v_rccallv(struct v_cstate *c, v_reg_type r) {
     v_calls++;
     v_rmach_call(c, r);
}

inline v_reg_type v_rccalli(struct v_cstate *c, v_reg_type r) {
     v_calls++;
     v_rmach_call(c, r);
     return v_caller_int_rr;
}

v_reg_type v_rccallu(struct v_cstate *c, v_reg_type r) {
     return v_rccalli(c, r);
}

v_reg_type v_rccallp(struct v_cstate *c, v_reg_type r) {
     return v_rccalli(c, r);
}

v_reg_type v_rccalll(struct v_cstate *c, v_reg_type r) {
     return v_rccalli(c, r);
}

v_reg_type v_rccallul(struct v_cstate *c, v_reg_type r) {
     return v_rccalli(c, r);
}

inline v_reg_type v_rccallf(struct v_cstate *c, v_reg_type r) {
     v_calls++;
     v_rmach_call(c, r);
     return v_fp_rr;
}

v_reg_type v_rccalld(struct v_cstate *c, v_reg_type r) {
     return v_rccallf(c, r);
}
