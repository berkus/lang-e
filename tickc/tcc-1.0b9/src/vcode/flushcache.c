#include <assert.h>
#include <stdio.h>

#include "vcode-internal.h"
#include "demand.h"

void v_flushcache(void *ptr, int nbytes) {
	int i;

	asm volatile ("iflush %0" : /* no outputs */ : "r" (ptr));

	/* make sure 8 byte aligned (I'm not sure this is necessary) */
	nbytes = v_roundup(nbytes, 8);

	/* Nuke every 8 bytes.   */
	for(i = 0; i < nbytes; i+=8, ptr = (char *)ptr + 8)
		asm volatile ("iflush %0" : /* no outputs */ : "r" (ptr));

	/* page 138: must be 5 instructions */
	asm volatile("nop");	/* 1 */
	asm volatile("nop");	/* 2 */
	asm volatile("nop");	/* 3 */
	asm volatile("nop");	/* 4 */
	asm volatile("nop");	/* 5 */
}
