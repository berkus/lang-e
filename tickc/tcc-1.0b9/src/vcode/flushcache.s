gcc2_compiled.:
___gnu_compiled_c:
.text
	.align 4
	.global _v_flushcache
	.proc	020
_v_flushcache:
	!#PROLOGUE# 0
	!#PROLOGUE# 1
	sll %o1,2,%g2
	iflush %o0
	add %g2,7,%g2
	and %g2,-8,%g2
	mov 0,%g3
	cmp %g3,%g2
	bge L3
	nop
L5:
	iflush %o0
	add %g3,4,%g3
	cmp %g3,%g2
	bl L5
	add %o0,4,%o0
L3:
	nop
	nop
	nop
	nop
	nop
	retl
	nop
