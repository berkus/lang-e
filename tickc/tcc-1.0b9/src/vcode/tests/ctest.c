/*
 * Test that calls with pushes and immediates work.  Also tests that 
 * vcode catches uninitialized cstates.
 */
#include <assert.h>
#include <stdio.h>
#include "vcode.h"

int main() { 
	static unsigned insn[1000];
	v_reg_type r;
	struct v_cstate c;
	v_vptr vp;

	v_clambda("foo", V_NLEAF, insn, sizeof insn);
	v_getreg(&r, V_I, V_TEMP);

	v_push_init(&c);
	v_push_argpi(&c, "Hello: %d %d %d %d\n");
	v_seti(r, 10);
	v_push_argi(&c, r);
	v_seti(r,20);
	v_push_argi(&c, r);
	v_seti(r,30);
	v_push_argi(&c, r);
	v_seti(r,40);
	v_push_argi(&c, r);
	v_ccallv(&c, (v_vptr)printf);
	vp = v_end(0).v;
#if 0
	v_dump((void*)vp);
#endif
	vp();

	v_clambda("foo", V_NLEAF, insn, sizeof insn);
	v_push_init(&c);
	v_push_argpi(&c, "Hello: %d %d %d %d\n");
	v_push_argii(&c, 10);
	v_push_argii(&c, 20);
	v_push_argii(&c, 30);
	v_push_argii(&c, 40);
	v_ccallv(&c, (v_vptr)printf);
	vp = v_end(0).v;
#if 0
	v_dump((void*)vp);
#endif
	vp();

#if 0
	v_clambda("foo", V_NLEAF, insn, sizeof insn);
	v_push_argpi(&c, "Hello: %d %d %d %d\n");
#endif
	return 0;
}
