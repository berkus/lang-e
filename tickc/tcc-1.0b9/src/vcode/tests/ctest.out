Hello: 10 20 30 40
Hello: 10 20 30 40
